### PR TITLE
[cmake] make TCP configurable with `OT_TCP` option

### PR DESCRIPTION
**Pull request by edmont**
Allow to remove TCP from the CMake build.

### CODE REVIEW

File: etc/cmake/options.cmake [hunk 1]
```diff
@@ -381,6 +381,13 @@ else()
     message(FATAL_ERROR "Invalid max RCP restoration count: ${OT_RCP_RESTORATION_MAX_COUNT}")
 endif()
 
+option(OT_TCP "enable TCPlp support")
+if(OT_TCP)
+    target_compile_definitions(ot-config INTERFACE "OPENTHREAD_CONFIG_TCP_ENABLE=1")
+else()
+    target_compile_definitions(ot-config INTERFACE "OPENTHREAD_CONFIG_TCP_ENABLE=0")
```
Comment (abtink) [inline etc/cmake/options.cmake:388]: While I think such a cmake option works and would exlcude the `tcplp` library from being included, it indirectly adds a restriction in that `TCP` under cmake can now only be enabled using this option.

In OT we have different ways of enabling features and providing the `OPENTHREAD_CONFIG_{}` definitions.
- We can use the build specific option, like cmake options above.
- Or a project can provide a project-specific config header, which includes all `OPENTHREAD_CONFIG_{}` definitions it wants and pass it as config header (e.g. in cmake using `OT_CONFIG`).

The model typically used for other cmake options is that they can enable a feature (set the related `OPENTHREAD_CONFIG_{}` to `1`), but won't explicitly disable it (wont explicitly set it to `0`). 

This way if the cmake option the is not specified/used, it allows the project specific config header to be used and specify the features to enable. 

----

I understand the main goal is to restrict inclusion of `tcplp` in `target_link_libraries`. Here is my suggestion for one possible way to realize this:
- We can add a new cmake option, `OT_EXCLUDE_TCPLP_LIB`
  - If it is used, we then exlude the `tcplp` including in link libs.
  - By default it is not set and therefore we include `tcplp`

By adding a different option for this (and not use `OT_TCP`) we keep these two independent of each other which allow a project-specific config header to be used under `cmake`. 

Thoughts?

**Comment by samkumar** [comment]
Overall, I am supportive of this. If the concern is increased code size, there may be flag we can add to the linker to remove unused symbols. But if the concern is increased build time, then I agree that a solution like this would be necessary.

@abtink is more familiar with OpenThread in general and its build system than I am, so I will defer to him for guidance on how best to handle this in he build system and in configuration files.

**Review by samkumar — APPROVED** [review]
Overall, I am supportive of this. If the concern is increased code size, there may be flag we can add to the linker to remove unused symbols. But if increased build times are also part of the concern, then I agree that a solution like this would be necessary.

@abtink is more familiar with OpenThread in general and its build system than I am, so I will defer to him for guidance on how best to handle this in the build system and in configuration files.

File: etc/cmake/options.cmake (hunk 1 above)
Comment (edmont) [reply etc/cmake/options.cmake]: Thanks for the clarification @abtink. I updated the PR with the suggested `OT_EXCLUDE_TCPLP_LIB` option.

BTW, it seems `OT_ASSERT` doesn't follow the general rule and has to match the `OPENTHREAD_CONFIG_ASSERT_ENABLE` value. Or am I wrong here?

Comment (abtink) [reply etc/cmake/options.cmake]: Thanks.

Yes. I noticed the same. Some features seem to set the related `OPENTHREAD_CONFIG` explicitly to zero from cmake option. I will submit a PR to change this (soon).

**Review by abtink — APPROVED** [review]
👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**